### DISCUSSION TITLE
질문

### DISCUSSION THREAD

**Discussion by Hsg2311** [MyeongWoonJang/cpp-study#93] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/75879925/200873479-07093e67-841a-4f4b-96d6-d1d60f9c8874.png)
ostream 자료형도 생소하고, ostream&처럼 return by reference 형식도 생소해서 이 코드를 이해하고 작성한 것이 아니라 외워서 작성한 느낌이 듭니다.

![image](https://user-images.githubusercontent.com/75879925/200874296-8a2ceb65-258f-452e-8fd4-32337df82cbf.png)
그리고 할당연산자들은 return *this를 해주는데 그냥 this를 return하면 안되는 건가요?

**Comment by Woon-2** [comment, marked as answer]
반환값에 대한 연계(`Method Chaining, 메서드 체이닝`)를 고려하셔야 합니다.   
어떤 함수의 반환값으로 다른 함수를 호출할 가능성이요.

`a = b`라는 표현은 사실 `a.operator=(b)`와 등가인데요.

`operator=`이라는 연산자가 좌변 객체에 대한 레퍼런스를 반환하지 않는다면
`a = b = c = d`와 같은 표현이 우리의 예상과는 다르게 흘러갑니다.   

등가 표현인 `a.operator=(b.operator=(c.operator=(d)))`에서,   
`c.operator=(d)`의 결과로 `c`가 튀어나오지 않는다는 소리와 같아요.   

`b.operator=(...)`의 `...`가 `c`여야 하는데, `c`가 아니라니요.   
`b`에 `c`가 대입되지 않는다면 도대체 뭐가 대입되나요?

`operator=`을 아예 반환값이 없는 `void` 함수로 작성했을 경우에는 컴파일조차 안되겠죠.   
`c.operator=(d)`의 결과가 없는데, 무(無)를 어떻게 `b`에 대입할 수 있을까요.

**연산자(`operator`)를 `overload`하는 것은, 연산자에 관한 새로운 문법을 만들겠다는 의지를 보이시는 겁니다.**      
그 문법이 어떻게 동작해야 하는지 원래는 언어 차원에서 정의하던 걸 여러분이 정의하는 겁니다.

여러분이 알고 있는 모든 연산자(`+`, `-`, `+=`, `-=`부터 시작해서 `->`, `[]`, `*(포인터 역참조)`, `<<`, `>>`까지)는 `overload`가 가능합니다.

<br>

위에선 연산자를 멤버 함수로 표현하였죠.

이항 연산자를 매개변수가 한 개인 멤버 함수로 정의할 수 있는 것은,   
멤버 함수의 경우

- 그 클래스 타입의 포인터가 암시적인 첫 번째 매개변수가 되고,
- 그 매개변수의 암시적인 인자로 `this`가 들어가기 때문입니다.

한 마디로,
`foo.do_something(bar);`라는 문장은
`do_something(&foo, bar);`와 같다는 뜻입니다.

"멤버 함수를 호출한 객체 자기 자신의 주소"가 항상 암시적인 첫 번째 인자가 된다구요.   
그것이 바로 `this`의 정체이고요.

다음 스프린트에서 클래스 문법 자체를 실제로 구현해보면서 설명하려 했었는데,   
개괄적인 내용은 위와 같습니다.

<br>

연산자 중에서는 비멤버 함수로 `overload`되는 것들도 있습니다.   

질문자분이 작성하신   
`std::ostream& operator<<(std::ostream& os, const Dog& dog)`의 경우   
대표적인 비멤버 이항 연산자이지요.

첫 번째 인자가 좌변, 두 번째 인자가 우변이 됩니다.   
`operator<<(os, dog)`과 `os << dog`이 등가란 소리입니다.   

이 친구도 역시   
`os << dog.name << ", " << dog.speed`와 같이 연계되고 있네요.   
`operator<<(operator<<(operator<<(os, dog.name), ", "), dog.speed)`와 등가입니다.   
괄호의 가장 안쪽부터 살펴보시면 납득이 가실 겁니다.

`std::ostream` 타입의 대표적인 객체로 `cout`이 있죠.   
생소해서 어렵다면, 그냥 `os` 자리에 `cout`이 있다고 생각해보세요.

<br>

반환 타입이 레퍼런스인 이유,   
`*this`를 반환하는 이유, 어쩌면 깨달으셨을 수도 있습니다.   

객체를 가지고 연계를 구성했는데 갑자기 객체의 주소가 튀어나오면 이상하잖아요?

다시 말하지만 여러분이 직접 문법을 구현하시는 겁니다.   

반환값으로 `->`나 `*`를 직접 찍어줘야 하는 불상사를 막으시려면,   
포인터인 `this`가 아니라 레퍼런스인 `*this`를 반환해야 하는 겁니다.
